### ISSUE TITLE
PYI019 causes false positives on `.py` files if you support Python <=3.10

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ruff#9761]
I just tried enabling ruff's PYI rules on [typeshed-stats](https://github.com/AlexWaygood/typeshed-stats), a hobby project of mine, and ran into a false positive with PYI019. I have the following class in the project:

```py
from enum import Enum
from typing import TypeVar

_NiceReprEnumSelf = TypeVar("_NiceReprEnumSelf", bound="_NiceReprEnum")

class _NiceReprEnum(Enum):
    """Base class for several public-API enums in this package."""

    def __new__(cls: type[_NiceReprEnumSelf], doc: str) -> _NiceReprEnumSelf:
        assert isinstance(doc, str)
        member = object.__new__(cls)
        member._value_ = member.__doc__ = doc
        return member
```

PYI019 tells me off about this, saying:

```
src\typeshed_stats\gather.py:88:60: PYI019 Methods like `__new__` should return `typing.Self` instead of a custom `TypeVar`
```

But I still support Python 3.10 (`typing.Self` is new in Python 3.11), and don't have `typing_extensions` as a dependency, so this violation isn't actionable.

(This isn't an issue for `.pyi` files, as type checkers think `typing_extensions` is part of the stdlib due to some white lies we tell them in typeshed. That means `Self` can always be imported from `typing_extensions` in a `.pyi` file, even if the stubs package doesn't explicitly declare a dependency on `typing_extensions`.)

**Comment by Bibo-Joshi** [comment]
Hi. I'd like to bump this issue again.
I see that #14230 by @charliermarsh addressed this issue and that PR was included in [v0.7.4](https://github.com/astral-sh/ruff/releases/tag/0.7.4). Indeed, running that version doesn't suggest using `typing.Self` for me. However, it seems to be back in the latest version, i.e. a regression:

foo.py
```python
from typing import TypeVar

Self = TypeVar("Self", bound="Foo")

class Foo:
    def return_self(self: Self) -> Self:
        ...
        return self
```

```shell
$ ruff --version
ruff 0.11.4
$ ruff check .\foo.py --isolated --target-version py39 --select PYI
foo.py:6:20: PYI019 Use `Self` instead of custom TypeVar `Self`
  |
5 | class Foo:
6 |     def return_self(self: Self) -> Self:
  |                    ^^^^^^^^^^^^^^^^^^^^ PYI019
7 |         ...
8 |         return self
  |
  = help: Replace TypeVar `Self` with `Self`

Found 1 error.
```

Am I missing something or is this indeed a regression?
Looking forward to feedback :)

**Comment by MichaReiser** [comment]
@Bibo-Joshi the error message here is a bit confusing because you named your `TypeVar` Self. But the rule recommends you to use `typing.Self` over your own custom `TypeVar`, see https://peps.python.org/pep-0673/

If you still think this isn't working as expected, please open a new issue

https://play.ruff.rs/ef57ccbf-05d1-4e54-9c6c-3e2365da3d5b

**Comment by Bibo-Joshi** [comment]
Hi, thanks for the quick reply. However I do not think that the problem is the name of the `TypeVar`. Renaming to `T` still gives the issue:

https://play.ruff.rs/8c7fc9c8-af0e-42fa-87ac-b3b9c86ede79

Or did I misunderstand your point? Please let me know if I should open a separate issue.

**Comment by AlexWaygood** [comment]
> Edit: Oh, I think the challenge here is that we have no good way of knowing whether a project has `typing_extensions` as a dependency or not.

yeah. And we have lots of rules that assume that you have `typing_extensions` as a dependency. Whatever we do here, we should fix it for all rules that have this problem at once, rather than applying a fix just to this rule.

Assuming you have `typing_extensions` as a dependency isn't an unreasonable assumption! Many, many projects have it either as an implicit or explicit dependency, since it is depended on by many core libraries and frameworks. So I wouldn't _necessarily_ want to disable this rule, and other rules like it, on lower Python versions.

It would be ideal if we could automatically detect whether or not the user has it listed as a dependency. Failing that, maybe we could consider a dedicated config option that you could toggle to tell Ruff whether you have it listed as a dependency, and use in rules like this...?

**Comment by Bibo-Joshi** [comment]
I see how parsing the dependencies can be difficult with all those different options out there 😁 introducing a config option to override the auto-detect sounds sane to me at First glance, though I have no idea in what ruffs policy on adding new configs is.
What I'd like to point out is that having `typing_extensions` as _implicit_ dependency should _not_ count towards this auto-detection IMO, as that's bad practice in the first place.

**Comment by Avasam** [comment]
I always avoid using `typing_extensions` as a runtime dependency unless I absolutely have to. But it can still be used for annotations no? (unless you need it at runtime).

```
from __future__ import annotations

from enum import Enum
from typing import TypeVar

if TYPE_CHECKING:
    from typing_extensions import Self

class _NiceReprEnum(Enum):
    """Base class for several public-API enums in this package."""

    def __new__(cls: type[_NiceReprEnumSelf], doc: str) -> Self:
        assert isinstance(doc, str)
        member = object.__new__(cls)
        member._value_ = member.__doc__ = doc
        return member
```

**Comment by Bibo-Joshi** [comment]
Typing-only dependencies are an interesting edge case, true. And they make an auto-detection of `typing_extensions` presence even harder. Still, I would also not like for ruff to throw an error telling me to switch to `typing_extensions` instead of using my custom `TypeVar` 😅

**Comment by AlexWaygood** [comment]
Yeah, I agree with @Bibo-Joshi -- this rule is purely stylistic... I'm not sure the awkwardness of having to add an `if TYPE_CHECKING` block for the typing-only import is really better stylistically than defining a custom TypeVar! Though if you already have an `if TYPE_CHECKING` block in your module, that's perhaps a different case

**Comment by MichaReiser** [comment]
A good first step would be to at least make the fix unsafe for `py` files when targeting <= 3.10 to avoid breaking tests
